### PR TITLE
Resolve and Unresolve

### PR DESCRIPTION
**Pull request by benibenj**
- Review Threads can be resolved/unresolved from the PullRequest Overview.
- ReviewThreads can be hidden.

### CODE REVIEW

File: src/github/pullRequestModel.ts [hunk 1]
```diff
@@ -997,6 +997,20 @@ export class PullRequestModel extends IssueModel<PullRequest> implements IPullRe
...
+			const prReviewThreadEvent = reviewEventsById[thread.prReviewDatabaseId];
+			prReviewThreadEvent.reviewThread = {
+				threadId: thread.id,
+				canResolve: thread.viewerCanResolve,
+				canUnresolve: thread.viewerCanUnresolve,
+				isResolved: thread.isResolved
+			};
+
+		});
+
```
Comment (alexr00) [inline src/github/pullRequestModel.ts:1013]: Why is this needed? The `PullRequestComments` query, which is what `reviewThreads` seems to come from, looks like it should have these properties already:

https://github.com/microsoft/vscode-pull-request-github/blob/da74def171b2bd0ff0123e015c3a67fb7fbe4ba8/src/github/queries.gql#L378-L386

File: webviews/common/context.tsx [hunk 2]
```diff
@@ -176,6 +176,13 @@ export class PRContext {
 
 	public openDiff = (comment: IComment) => this.postMessage({ command: 'pr.open-diff', args: { comment } });
 
+	public toggleResolveComment = (reviewThread: ReviewEvent['reviewThread'], thread: IComment[]) => {
+		if (!reviewThread) {
+			return;
+		}
+		this.postMessage({ command: 'pr.resolve-comment-thread', args: { threadId: reviewThread.threadId, toResolve: reviewThread.isResolved, thread } });
```
Comment (alexr00) [inline webviews/common/context.tsx:183]: Might need to await this `postMessage` and then call `updatePR`. That could be where the unexpected UI behavior we were seeing is coming from.

File: src/github/pullRequestModel.ts (hunk 1 above)
Comment (benibenj) [reply src/github/pullRequestModel.ts:1013]: We are getting the information from `PullRequestComments`, but need the TimelineEvent to know about this information. And as the TimelineEvents API does not know about the threadId and resolve permissions, we need to add this information to the event.

File: webviews/common/context.tsx (hunk 2 above)
Comment (benibenj) [reply webviews/common/context.tsx]: Actually I was only missing the `updatePR` call. I was changing the values in the timeline but I did not update the state which caused the inconsistent state.

File: webviews/common/context.tsx [hunk 3]
```diff
@@ -176,6 +176,17 @@ export class PRContext {
...
 	public openDiff = (comment: IComment) => this.postMessage({ command: 'pr.open-diff', args: { comment } });
 
+	public toggleResolveComment = (reviewThread: ReviewEvent['reviewThread'], thread: IComment[], newResolved: boolean) => {
+		if (!reviewThread) {
+			return;
+		}
+		reviewThread.isResolved = newResolved;
+		reviewThread.canUnresolve = newResolved;
+		reviewThread.canResolve = !newResolved;
+		this.updatePR({});
```
Comment (alexr00) [inline webviews/common/context.tsx:186]: The usual patter for `updatePR` is to pass the properteis of the PR that changed into `updatePR`. Does just calling it without passing in the updated `{ events }` work?

Comment (alexr00) [reply webviews/common/context.tsx:186]: I guess you're modifying the original object, so that's why it would work. I find this to be a little tricky to understand and not self-documenting (I have to make sure I understand that `reviewThread` is the same object that is already in `this.pr`, there are no clues about what is happening). Instead, is it possible to use the usual `const response = await this.postMessage` then pass some property of the `response` into `updatePR`?

Comment (benibenj) [reply webviews/common/context.tsx:186]: It seemed to work but I just noticed it actually doe not. I'm trying to figure out how to only update the state of one array item

File: webviews/common/context.tsx [hunk 4]
```diff
@@ -176,6 +176,17 @@ export class PRContext {
 
 	public openDiff = (comment: IComment) => this.postMessage({ command: 'pr.open-diff', args: { comment } });
 
+	public toggleResolveComment = (reviewThread: ReviewEvent['reviewThread'], thread: IComment[], newResolved: boolean) => {
```
Comment (alexr00) [inline webviews/common/context.tsx:179]: nit: I would prefer we break `ReviewThread` into a separate type.

File: src/common/timelineEvent.ts [hunk 5]
```diff
@@ -39,8 +39,16 @@ export interface CommentEvent {
 	createdAt: string;
 }
 
+export interface reviewResolveInfo {
```
Comment (alexr00) [inline src/common/timelineEvent.ts:42]: The style in this codebase is that interface names are Pascal case.

Comment (alexr00) [reply src/common/timelineEvent.ts:42]: And the signature [here](https://github.com/microsoft/vscode-pull-request-github/blob/9859e75b24f99d8402fc443a7bd286e6ed0bc1c5/webviews/common/context.tsx#L179) should use the new type.

File: webviews/common/context.tsx (hunk 3 above)
Comment (alexr00) [reply webviews/common/context.tsx:186]: Would it make sense to call `PullRequestModel`'s `getTimelineEvents` method after doing the resolve in `PullRequestOverViewPanel`? That would also have the benefit of updating state that the overview uses so that we're up-to-date in both the extension and the webview as well as giving you the correct item to include in the message response.

**Review by alexr00 — CHANGES_REQUESTED** [review]

File: src/github/pullRequestModel.ts [hunk 6]
```diff
@@ -938,17 +939,25 @@ export class PullRequestModel extends IssueModel<PullRequest> implements IPullRe
...
-			const { data } = await query<TimelineEventsResponse>({
-				query: schema.TimelineEvents,
-				variables: {
-					owner: remote.owner,
-					name: remote.repositoryName,
-					number: this.number,
-				},
-			});
+			const [{ data }, latestReviewCommitInfo, currentUser, reviewThreads] = await Promise.all([
+				await query<TimelineEventsResponse>({
```
Comment (alexr00) [inline src/github/pullRequestModel.ts:943]: unneeded await?

File: webviews/common/context.tsx [hunk 7]
```diff
@@ -5,7 +5,7 @@
 
 import { createContext } from 'react';
 import { IComment } from '../../src/common/comment';
-import { EventType, isReviewEvent, ReviewEvent } from '../../src/common/timelineEvent';
+import { EventType, isReviewEvent, ReviewEvent, ReviewResolveInfo, TimelineEvent } from '../../src/common/timelineEvent';
```
Comment (alexr00) [inline webviews/common/context.tsx:8]: lint error with unused `ReviewResolveInfo`. I'll try to see why the build isn't catching this.

**Review by alexr00 — DISMISSED** [review]

File: src/github/pullRequestModel.ts (hunk 6 above)
Comment (benibenj) [reply src/github/pullRequestModel.ts:943]: yes, thanks for noticing

**Review by alexr00 — APPROVED** [review]
👍
